### PR TITLE
emit metrics for informer watch errors

### PR DESCRIPTION
**Pull request by stevenctl**
related to #30838 – we need visibility into informer errors, which controller they're occurring for (this wires it up for serviceentry registry and k8s registry)

### CODE REVIEW

File: pilot/pkg/controller/informerutil.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	controller string
+	lastErr    *atomic.Error
+	errors     monitoring.Metric
+	recoveries monitoring.Metric
+}
+
+func (h *informerErrHandler) clearError(reason string) {
+	if h.lastErr.Load() != nil {
+		h.lastErr.Store(nil)
+		h.recoveries.Increment()
```
Comment (stevenctl) [inline pilot/pkg/controller/informerutil.go:44]: I'm actually not sure if this is needed.

File: pilot/pkg/config/kube/ingressv1/controller.go [hunk 2]
```diff
@@ -237,6 +238,17 @@ func (c *controller) RegisterEventHandler(kind config.GroupVersionKind, f func(c
 	}
 }
 
+func (c *controller) SetWatchErrorHandler(handler func(r *cache.Reflector, err error)) error {
+	var errs error
+	if err := c.serviceInformer.SetWatchErrorHandler(handler); err != nil {
```
Comment (howardjohn) [inline pilot/pkg/config/kube/ingressv1/controller.go:243]: So this serviceInformer is actually shared by other places - does this handle it ok? IIRC its "last one wins" which is fine if they always use the same handler?

Comment (stevenctl) [reply pilot/pkg/config/kube/ingressv1/controller.go:243]: ah, I was hoping that the metric included a label indicating which component was using the informer, that part won't be accurate – the more important part is the cluster label

I'll trim this down a bit

File: pilot/pkg/util/informermetric/informerutil.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+	handler, ok := handlers[clusterID]
+	mu.RUnlock()
+	if ok {
+		return handler
+	}
+
+	mu.Lock()
+	defer mu.Unlock()
+	metric := errorMetric.With(clusterLabel.Value(clusterID))
+	if err := metric.Register(); err != nil {
```
Comment (stevenctl) [inline pilot/pkg/util/informermetric/informerutil.go:51]: Do I need to register each variant with the cluster label? 
If so, can I register a "root" metric with the label and no value?

Comment (bianpengyuan) [reply pilot/pkg/util/informermetric/informerutil.go:51]: Yeah you just need to register the metric it self, likely in an init function. No need to register every variant of time series.

Comment (stevenctl) [reply pilot/pkg/util/informermetric/informerutil.go:51]: ty!

**Review by ericvn — APPROVED** [review]
Release note looks good

File: pilot/pkg/util/informermetric/informerutil.go [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+	"istio.io/pkg/log"
+	"istio.io/pkg/monitoring"
+)
+
+var (
+	clusterLabel = monitoring.MustCreateLabel("cluster")
+
+	errorMetric = monitoring.NewSum(
+		"controller_sync_errors",
```
Comment (howardjohn) [inline pilot/pkg/util/informermetric/informerutil.go:30]:
```suggestion
		"controller_sync_errors_total",
```
per https://prometheus.io/docs/practices/naming/#metric-names

**Review by howardjohn — APPROVED** [review]

**Pull request merged**